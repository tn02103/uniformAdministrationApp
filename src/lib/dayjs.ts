import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
import customParseFormat from "dayjs/plugin/customParseFormat";
import timezone from "dayjs/plugin/timezone";

dayjs.extend(utc);
dayjs.extend(customParseFormat);
dayjs.extend(timezone);
dayjs.tz.setDefault("Germany/Berlin");

import "dayjs/locale/de";
import "dayjs/locale/en";

export default dayjs;