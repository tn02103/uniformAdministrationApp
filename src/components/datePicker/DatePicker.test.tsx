import dayjs from "@/lib/dayjs";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useState } from "react";
import { DatePicker, DatePickerProps, INVALID_DATE } from "./DatePicker";

describe('DatePicker', () => {

    const DatePickerTestField = (props: Partial<DatePickerProps>) => {
        const [value, setValue] = useState<string | null | typeof INVALID_DATE>(null);
        const handleChange = (date: string | null | typeof INVALID_DATE) => {
            setValue(date);
            props.onChange?.(date);
        }

        return (
            <DatePicker
                value={value}
                {...props}
                onChange={handleChange}
            />
        );
    };

    it('should render without crashing', () => {
        const onChange = jest.fn();
        const { container } = render(<DatePicker onChange={onChange} value={null} />);
        expect(screen.getByRole('textbox')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should open and close calendar', async () => {
        const user = userEvent.setup();
        const onChange = jest.fn();
        render(<DatePickerTestField onChange={onChange} />);

        const button = screen.getByRole('button', { name: /calendar/i });
        await user.click(button);

        expect(screen.getByRole('dialog')).toBeInTheDocument();

        await user.click(screen.getByRole('button', { name: /15/ }));
        expect(screen.queryByRole('dialog')).toBeNull();

        await user.click(button);
        expect(screen.getByRole('dialog')).toBeInTheDocument();

        await user.click(screen.getByRole('textbox'));
        expect(screen.queryByRole('dialog')).toBeNull();
    });

    it('should handle input change', async () => {
        const user = userEvent.setup();
        const onChange = jest.fn();
        render(<DatePickerTestField onChange={onChange} />);

        const input = screen.getByRole('textbox');
        await user.type(input, '01.01');
        expect(input).toHaveValue('01.01');
        expect(onChange).toHaveBeenLastCalledWith(INVALID_DATE);

        await user.type(input, '.2020');
        expect(input).toHaveValue('01.01.2020');

        expect(onChange).toHaveBeenLastCalledWith('2020-01-01');

        await user.clear(input);
        expect(input).toHaveValue('');
        expect(onChange).toHaveBeenLastCalledWith(null);
    });

    it('should handle select via calendar', async () => {
        const user = userEvent.setup();
        const onChange = jest.fn();
        render(<DatePickerTestField onChange={onChange} />);

        const button = screen.getByRole('button', { name: /calendar/i });
        await user.click(button);

        const dayButton = screen.getByText('15');
        await user.click(dayButton);

        const expectedDate = dayjs().date(15);
        expect(onChange).toHaveBeenLastCalledWith(expectedDate.format('YYYY-MM-DD'));
        expect(screen.getByRole('textbox')).toHaveValue(expectedDate.format('DD.MM.YYYY'));
    });

    it('should handle minDate', async () => {
        const user = userEvent.setup();
        const onChange = jest.fn();
        const minDate = dayjs().date(15);
        render(<DatePickerTestField onChange={onChange} minDate={minDate.toDate()} />);

        const button = screen.getByRole('button', { name: /calendar/i });
        await user.click(button);

        const prevDayButton = screen.getByRole('button', { name: /14/ });
        expect(prevDayButton).toBeDisabled();
        await user.click(prevDayButton);
        expect(onChange).not.toHaveBeenCalled();

        const minDayButton = screen.getByRole('button', { name: /15/ });
        expect(minDayButton).toMatchSnapshot();
        await user.click(minDayButton);

        expect(onChange).toHaveBeenLastCalledWith(minDate.format('YYYY-MM-DD'));
        expect(screen.getByRole('textbox')).toHaveValue(minDate.format('DD.MM.YYYY'));
    });
});
