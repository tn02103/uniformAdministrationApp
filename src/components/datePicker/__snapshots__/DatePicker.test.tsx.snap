// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should handle minDate 1`] = `
<button
  class="react-calendar__tile react-calendar__month-view__days__day"
  style="flex-basis: 14.285714285714286%; flex-shrink: 0; flex-grow: 0; overflow: hidden; margin-inline-end: 0;"
  type="button"
>
  <abbr
    aria-label="May 15, 2025"
  >
    15
  </abbr>
</button>
`;

exports[`DatePicker should render without crashing 1`] = `
<div>
  <div
    class="position-relativ"
  >
    <div
      class="input-group flex"
    >
      <input
        class="form-control "
        name="date"
        placeholder="01.01.2020"
        type="string"
        value=""
      />
      <button
        aria-label="open calendar"
        class="input-group-text"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-calendar "
          data-icon="calendar"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32l0 32L48 64C21.5 64 0 85.5 0 112l0 48 448 0 0-48c0-26.5-21.5-48-48-48l-48 0 0-32c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 32L160 64l0-32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192L0 192 0 464c0 26.5 21.5 48 48 48l352 0c26.5 0 48-21.5 48-48l0-272z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      style="display: contents;"
    />
  </div>
</div>
`;
